refactor(config)!: Standardize TOML parsing using direct Serde deserialization

gustavovalverde opened the pull request:
## Description

This PR refactors the configuration parsing logic in `zaino/zainod/src/config.rs` to standardize it based on idiomatic Serde practices.

The key changes include:

-   **Elimination of Special "None" String Handling**: The previous logic that treated the string literal `"None"` as a way to unset optional path fields has been removed.
-   **Direct Serde Deserialization**:
    -   The `IndexerConfig` struct now directly uses `#[derive(serde::Deserialize)]`.
    -   The custom `load_config` function has been significantly simplified to use `toml::from_str` for parsing the entire configuration into `IndexerConfig`.
    -   The `parse_field_or_warn_and_default!` macro has been removed.
-   **Custom Deserializers**: For fields requiring specific string-to-type transformations (e.g., `SocketAddr` from a hostname, `BackendType` from "fetch" or "state"), `#[serde(deserialize_with = "...")]` is now used with dedicated deserialization functions.
-   **Post-Deserialization Logic**: Conditional default logic, such as determining `cookie_dir` based on `enable_cookie_auth`, is now handled in a new `finalize_config_logic` method within `IndexerConfig`, called after the initial Serde parsing.
-   **Error Handling**: If the TOML file cannot be read or parsed, `load_config` now logs a warning and falls back to `IndexerConfig::default()`, which is then also finalized and checked.
-   **Unit Tests**: A comprehensive suite of unit tests has been added to `config.rs` to validate:
    -   Parsing of various valid and invalid TOML inputs.
    -   Correct defaulting of optional fields.
    -   The new behavior of how literal `"None"` strings are handled for different `Option` types.
    -   Inclusion of the project's actual `zaino/zainod/zindexer.toml` (via `include_str!`) to confirm its parsing behavior (it will now error on fields like `map_capacity = "None"`).

This approach aligns Zaino's configuration handling more closely with practices seen in Zebra and Zallet, and standard Rust/Serde idioms, improving clarity and maintainability.

## Breaking Change

This change modifies how optional fields in `zindexer.toml` are interpreted:

1.  **For `Option<String>` or `Option<PathBuf>` fields**: If a field like `cookie_dir` or `tls_cert_path` is set to the string `"None"` in the TOML (e.g., `cookie_dir = "None"`), it will now be parsed as `Some(PathBuf::from("None"))` or `Some("None".to_string())` respectively. To have these fields be `std::option::Option::None`, the key should be entirely **omitted** from the TOML file.
    *   *Exception*: `cookie_dir` will still become `None` if `enable_cookie_auth` is `false`, due to the `finalize_config_logic` method.
2.  **For other `Option<T>` fields (e.g., `Option<usize>`, `Option<bool>`)**: Setting such a field to the string `"None"` (e.g., `map_capacity = "None"`) will now correctly result in a **TOML parsing error**. To have these fields be `std::option::Option::None`, the key must be **omitted** from the TOML file.

Users who previously relied on setting fields to the string `"None"` to indicate an unset optional value will need to update their `zindexer.toml` files by removing these lines instead.


Relates to:
- #331 
- ADR: https://github.com/zingolabs/zaino/pull/331/files#diff-e44d81a13db5cbf1c5e3ffffe53f0ff1206522bc72e3e8d4595d3d89897d107d

gustavovalverde committed:
refactor(config)!: Standardize TOML parsing using direct Serde deserialization

Removes custom field-by-field parsing and special handling of "None" strings
for optional paths in `zaino/zainod/src/config.rs`.

The configuration `IndexerConfig` now derives `serde::Deserialize` and
relies on `toml::from_str` for parsing. Custom deserialization functions
are used for specific types like `SocketAddr` and `BackendType`.
Conditional default logic (e.g., `cookie_dir`) is handled in a
post-deserialization `finalize_config_logic` method.

BREAKING CHANGE:
Optional fields in `zindexer.toml` that are not strings or paths (e.g.,
`Option<usize>`, `Option<bool>`) and were previously set to the string
literal "None" will now cause a parsing error. These keys should be
omitted from the TOML to be interpreted as `None`.
Optional string or path fields set to "None" will now be interpreted as
the literal string/path "None", rather than `std::option::Option::None`.
Omit these keys for them to be `None`.

idky137 committed:
Merge branch 'dev' into fix-toml-parsing

idky137 commented on zainod/src/config.rs:
```diff
@@ -611,3 +467,499 @@ impl TryFrom<IndexerConfig> for BackendConfig {
+
+        // Assert that parsing the original zindexer.toml content fails
+        // due to string "None" for Option<usize> fields (and similar issues).
+        assert!(
```
Should this assert be here? 

It looks like it was used to test that the old "zindexer.toml" would fail, however since this has been updated to the new format this test is now failing for me.

gustavovalverde replied:
Sorry about that, this was an oversight while I was testing with the previous file. Then I updated the `zindexer.toml` as there was no use in having it with the previous format, and I forgot to update the tests to reflect this.

I updated the test and ran a `cargo test -p zainod` to confirm, this is now fixed here
[test(config): Update zindexer.toml parsing test for new format](https://github.com/zingolabs/zaino/pull/334/commits/eb9cbcbad0071d4ded99b436ad982479ebc2a6ac)

idky137 reviewed (changes requested):
Thanks for this! Looks much nicer!

The only issue I can see is `test_parse_existing_zindexer_toml_content()` looking at the new format instead of the old.

gustavovalverde committed:
test(config): Update zindexer.toml parsing test for new format

Renames `test_parse_existing_zindexer_toml_content` to `test_parse_zindexer_toml`.

This test now asserts that the current `zaino/zainod/zindexer.toml`
(loaded via `include_str!`) parses successfully, reflecting updates
to the file to align with the new Serde-based parsing rules.

gustavovalverde commented:
@idky137 I'm closing this PR in favor of #335

That implementation is cleaner, the test are separated (to avoid cluttering the module logic) and it avoid implementing a custom docker entrypoint with all the variables consideration as we have in Zebra https://github.com/ZcashFoundation/zebra/blob/main/docker/entrypoint.sh#L41-L130 and as I was trying to do here: https://github.com/zingolabs/zaino/pull/331/files#diff-4f5cabe26761257a4d685a6edc7a43e0fe0f78762f50eeb48530f2bd3b3ee7ca

gustavovalverde commented:
@idky137 I'm closing this PR in favor of #335

That implementation is cleaner, the test are separated (to avoid cluttering the module logic) and it avoid implementing a custom docker entrypoint with all the variables consideration as we have in Zebra https://github.com/ZcashFoundation/zebra/blob/main/docker/entrypoint.sh#L41-L130 and as I was trying to do here: https://github.com/zingolabs/zaino/pull/331/files#diff-4f5cabe26761257a4d685a6edc7a43e0fe0f78762f50eeb48530f2bd3b3ee7ca

gustavovalverde commented:
Hey @idky137, let me know if you have questions for #335
